### PR TITLE
feat: Add side input sdkclient and grpc 

### PR DESCRIPTION
**Pull request by kohlisid**

### CODE REVIEW

File: pkg/sideinputs/manager/manager.go [hunk 1]
```diff
@@ -52,31 +55,45 @@ func (sim *sideInputsManager) Start(ctx context.Context) error {
...
+	}
+	// close the connection when the function exits
+	defer func() {
+		err = sideInputClient.CloseConn(ctx)
+		if err != nil {
+			log.Warnw("Failed to close gRPC client conn", zap.Error(err))
+		}
+	}()
+
+	sideInputBucketName := isbsvc.JetStreamSideInputsStoreBucket(sim.sideInputsStore)
```
Comment (whynowy) [inline pkg/sideinputs/manager/manager.go:89]: This section should be moved to the place of line 64.

File: pkg/sideinputs/manager/manager.go [hunk 2]
```diff
@@ -96,10 +113,17 @@ func (sim *sideInputsManager) Start(ctx context.Context) error {
 	return nil
 }
 
-func (sim *sideInputsManager) execute(ctx context.Context) error {
+func (sim *sideInputsManager) execute(ctx context.Context, sideInputClient *udsideinput.UDSgRPCBasedUDSideinput, siStore kvs.KVStorer) error {
 	log := logging.FromContext(ctx)
-	// TODO(SI): call ud container to fetch data and write to store.
-	log.Info("Executing ...")
+	log.Info("Executing Side Input manager cron ...")
```
Comment (whynowy) [inline pkg/sideinputs/manager/manager.go:118]: `Side Inputs manager`

File: pkg/sideinputs/synchronizer/synchronizer.go [hunk 3]
```diff
@@ -104,13 +106,17 @@ func startSideInputSynchronizer(ctx context.Context, watch kvs.KVWatcher, mountP
...
-			log.Debug("Side Input value received ",
+			log.Infow("Side Input value received ",
 				zap.String("key", value.Key()), zap.String("value", string(value.Value())))
 			p := path.Join(mountPath, value.Key())
-			// Write changes to disk
-			err := utils.UpdateSideInputFile(ctx, p, value.Value())
-			if err != nil {
-				log.Errorw("Failed to update Side Input value %s", zap.Error(err))
+			// Check if the value has changed from the last update
+			if prevValue == nil || bytes.Equal(prevValue, value.Value()) == false {
```
Comment (whynowy) [inline pkg/sideinputs/synchronizer/synchronizer.go:113]: Do the check here means when a pod starts, there will be a consecutive file changes - one is done by the init container, the other one is done by this. Shall we do it in the function of `UpdateSideInputFile()`? To read the old file content and compare?

Comment (kohlisid) [reply pkg/sideinputs/synchronizer/synchronizer.go]: Done

**Comment by vigith** [comment]
tests are still failing

File: pkg/sideinputs/manager/manager.go [hunk 4]
```diff
@@ -52,31 +55,48 @@ func (sim *sideInputsManager) Start(ctx context.Context) error {
...
 	switch sim.isbSvcType {
 	case dfv1.ISBSvcTypeRedis:
 		return fmt.Errorf("unsupported isbsvc type %q", sim.isbSvcType)
 	case dfv1.ISBSvcTypeJetStream:
-		natsClient, err := jsclient.NewNATSClient(ctx)
+		natsClient, err = jsclient.NewNATSClient(ctx)
 		if err != nil {
 			log.Errorw("Failed to get a NATS client.", zap.Error(err))
 			return err
 		}
```
Comment (yhl25) [inline pkg/sideinputs/manager/manager.go:69]: client is being closed?

File: pkg/sdkclient/sideinput/client.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute grpc.Dial(%q): %w", sockAddr, err)
+	}
+	c.conn = conn
+	c.grpcClt = sideinputpb.NewSideInputClient(conn)
+	return c, nil
+}
+
+// NewFromClient creates a new client object from a grpc client. This is used for testing.
+func NewFromClient(c sideinputpb.SideInputClient) (Client, error) {
```
Comment (whynowy) [inline pkg/sdkclient/sideinput/client.go:47]: Is there a better way?

File: pkg/sideinputs/udsideinput/udsideinputs_grpc.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+	v1 "github.com/numaproj/numaflow-go/pkg/apis/proto/sideinput/v1"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/numaproj/numaflow/pkg/sdkclient/sideinput"
+)
+
+// UDSgRPCBasedUDSideinput applies user defined side input over gRPC (over Unix Domain Socket) client/server
+// where server is the UDSideInput.
+type UDSgRPCBasedUDSideinput struct {
```
[the remark this answers is not in the document]
Comment (kohlisid) [reply pkg/sideinputs/udsideinput/udsideinputs_grpc.go]: Using the client directly now

File: pkg/sideinputs/manager/manager.go [hunk 7]
```diff
@@ -52,31 +56,50 @@ func (sim *sideInputsManager) Start(ctx context.Context) error {
...
+	if err != nil {
+		return fmt.Errorf("failed to create a new gRPC client: %w", err)
+	}
+
+	// Readiness check
+	if err = sideInputClient.WaitUntilReady(ctx); err != nil {
+		return fmt.Errorf("failed on SideInput readiness check, %w", err)
+	}
+	// close the connection when the function exits
+	defer func() {
```
Comment (whynowy) [inline pkg/sideinputs/manager/manager.go:94]: This should be moved to line 88.

**Review by whynowy — APPROVED** [review]
Minor comments.

File: pkg/sideinputs/utils/utils.go [hunk 8]
```diff
@@ -40,8 +43,20 @@ func UpdateSideInputFile(ctx context.Context, fileSymLink string, value []byte)
 	timestamp := time.Now().UnixNano()
 	newFileName := fmt.Sprintf("%s_%d", fileSymLink, timestamp)
 
+	// Fetch the current side input value from the file
+	currentValue, err := FetchSideInputFileValue(fileSymLink)
+
+	// Check if the current value is same as the new value
+	// If true then don't update file again and return
+	if err == nil && bytes.Equal(currentValue, value) {
```
Comment (jy4096) [inline pkg/sideinputs/utils/utils.go:51]: in the future when file size gets larger we may consider using hash...

**Review by jy4096 — APPROVED** [review]

**Review by vigith — APPROVED** [review]

**Pull request merged**